File: src/extras-sonnet/libguesslanguage.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/sonnet-guesslanguage.html
pub const sonnet__guesslanguage = struct {
    /// New constructs a new Sonnet::GuessLanguage object.
    ///
    ///
    pub fn New() QtC.Sonnet__GuessLanguage {
        return qtc.Sonnet__GuessLanguage_new();
    }

    /// [Qt documentation](https://api.kde.org/sonnet-guesslanguage.html#setLimits)
    ///
    /// ``` self: QtC.Sonnet__GuessLanguage, maxItems: i32, minConfidence: f64 ```
    pub fn SetLimits(self: ?*anyopaque, maxItems: i32, minConfidence: f64) void {
        qtc.Sonnet__GuessLanguage_SetLimits(@ptrCast(self), @intCast(maxItems), @floatCast(minConfidence));
    }

    /// [Qt documentation](https://api.kde.org/sonnet-guesslanguage.html#identify)
    ///
    /// ``` self: QtC.Sonnet__GuessLanguage, text: []const u8, allocator: std.mem.Allocator ```
    pub fn Identify(self: ?*anyopaque, text: []const u8, allocator: std.mem.Allocator) []const u8 {
        const text_str = qtc.libqt_string{
            .len = text.len,
            .data = text.ptr,
        };
        const _str = qtc.Sonnet__GuessLanguage_Identify(@ptrCast(self), text_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("sonnet::guesslanguage.Identify: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/sonnet-guesslanguage.html#identify)
    ///
    /// ``` self: QtC.Sonnet__GuessLanguage, text: []const u8, suggestions: [][]const u8, allocator: std.mem.Allocator ```
    pub fn Identify2(self: ?*anyopaque, text: []const u8, suggestions: [][]const u8, allocator: std.mem.Allocator) []const u8 {
        const text_str = qtc.libqt_string{
            .len = text.len,
            .data = text.ptr,
        };
        var suggestions_arr = allocator.alloc(qtc.libqt_string, suggestions.len) catch @panic("sonnet::guesslanguage.Identify2: Memory allocation failed");
        defer allocator.free(suggestions_arr);
        for (suggestions, 0..suggestions.len) |item, i| {
            suggestions_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const suggestions_list = qtc.libqt_list{
            .len = suggestions.len,
            .data = suggestions_arr.ptr,
        };
        const _str = qtc.Sonnet__GuessLanguage_Identify2(@ptrCast(self), text_str, suggestions_list);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("sonnet::guesslanguage.Identify2: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Sonnet__GuessLanguage ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Sonnet__GuessLanguage_Delete(@ptrCast(self));
    }
};

File: src_original/extras-sonnet/libguesslanguage.cpp
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#define WORKAROUND_INNER_CLASS_DEFINITION_Sonnet__GuessLanguage
#include <guesslanguage.h>
#include "libguesslanguage.h"
#include "libguesslanguage.hxx"

Sonnet__GuessLanguage* Sonnet__GuessLanguage_new() {
    return new Sonnet::GuessLanguage();
}

void Sonnet__GuessLanguage_SetLimits(Sonnet__GuessLanguage* self, int maxItems, double minConfidence) {
    self->setLimits(static_cast<int>(maxItems), static_cast<double>(minConfidence));
}

libqt_string Sonnet__GuessLanguage_Identify(const Sonnet__GuessLanguage* self, const libqt_string text) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    QString _ret = self->identify(text_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string Sonnet__GuessLanguage_Identify2(const Sonnet__GuessLanguage* self, const libqt_string text, const libqt_list /* of libqt_string */ suggestions) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    QList<QString> suggestions_QList;
    suggestions_QList.reserve(suggestions.len);
    libqt_string* suggestions_arr = static_cast<libqt_string*>(suggestions.data);
    for (size_t i = 0; i < suggestions.len; ++i) {
        QString suggestions_arr_i_QString = QString::fromUtf8(suggestions_arr[i].data, suggestions_arr[i].len);
        suggestions_QList.push_back(suggestions_arr_i_QString);
    }
    QString _ret = self->identify(text_QString, suggestions_QList);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Sonnet__GuessLanguage_Delete(Sonnet__GuessLanguage* self) {
    delete self;
}
